py/obj.h

```diff
@@ -743,15 +743,30 @@ void *mp_obj_malloc_helper(size_t num_bytes, const mp_obj_type_t *type);
 // check for more specific object types.
 // Note: these are kept as macros because inline functions sometimes use much
 // more code space than the equivalent macros, depending on the compiler.
-#define mp_obj_is_type(o, t) (mp_obj_is_obj(o) && (((mp_obj_base_t *)MP_OBJ_TO_PTR(o))->type == (t))) // this does not work for checking int, str or fun; use below macros for that
+// don't use mp_obj_is_exact_type directly; use mp_obj_is_type which provides additional safety checks.
+// use the former only if you need to bypass these checks (because you've already checked everything else)
+#define mp_obj_is_exact_type(o, t) (mp_obj_is_obj(o) && (((mp_obj_base_t *)MP_OBJ_TO_PTR(o))->type == (t)))
+
+// type checks are split to a separate, constant result macro. I found this way to work without hindering GCC's
+// optimizations (other tricks like using ({ expr; exper; }) or (exp, expr, expr) in mp_obj_is_type result in
+// missed optimizations. however, this way GCC retains the optimizations it used to have before these checks
+// were added).
```

Comment:
Better to write comments not in the first person. And also no real need to refer to the state of the code before this change was made (because what matters moving forward is how the code is now).  Eg something like "Type checks are split to a separate, constant result macro.  This is so it doesn't hinder the compiler's optimizations (other tricks like ... result in missed optimizations)."

I'm happy to make the change myself during merge.  Up to you.

Reply:
I'll make those changes now

Reply:
Done :)